### PR TITLE
Convert uses of the Dangerous APIs to use MemoryMarshal.GetReference

### PR DESCRIPTION
**Pull request by ahsonkhan**
Part of:
https://github.com/dotnet/corefx/issues/25412
https://github.com/dotnet/corefx/issues/25615

Related to (and depends on): https://github.com/dotnet/coreclr/pull/15532

**Note:** Expecting this failure for now until we have resolved whether we should add `Unsafe.AsRef<T>(in T)` or not. See: https://github.com/dotnet/coreclr/pull/15532#issuecomment-351896709
```text
Cannot use method 'MemoryMarshal.GetReference<char>(ReadOnlySpan<char>)' as a ref or out value because it is a readonly variable
```

Following the staging plan from here: https://github.com/dotnet/corefx/issues/23881#issuecomment-343767740

- [x] Add MemoryExtensions.GetReference/TryGetArray
- [ ] Convert all uses of DangerousGetPinnableReference/DangerousTryGetArray in coreclr, corefx, corert, corefxlab, aspnet, ... to MemoryExtensions.GetReference
- [ ] Change DangerousGetPinnableReference to whatever we like to make it fit the pinning pattern and remove DangerousTryGetArray.

Doing it this way will avoid the need for complex staging or things being on the floor for extensive periods of time.

cc @jkotas, @stephentoub, @KrzysztofCwalina, @davidfowl, @pakrym, @joshfree

### CODE REVIEW

File: src/System.Security.Cryptography.Algorithms/src/System.Security.Cryptography.Algorithms.csproj [hunk 1]
```diff
@@ -481,6 +481,7 @@
     <Reference Include="System.Diagnostics.Contracts" />
     <Reference Include="System.Diagnostics.Debug" />
     <Reference Include="System.Diagnostics.Tools" />
+    <Reference Include="System.Memory" />
```
Comment (KrzysztofCwalina) [inline src/System.Security.Cryptography.Algorithms/src/System.Security.Cryptography.Algorithms.csproj]: Why did you have to add these?

Comment (ahsonkhan) [reply src/System.Security.Cryptography.Algorithms/src/System.Security.Cryptography.Algorithms.csproj]: MemoryMarshal is available through System.Memory.

File: src/System.IO.FileSystem/src/System/IO/PathHelpers.cs [hunk 2]
```diff
@@ -136,6 +137,14 @@ private unsafe static string CombineNoChecksInternal(ReadOnlySpan<char> first, R
         private unsafe static string CombineNoChecksInternal(ReadOnlySpan<char> first, ReadOnlySpan<char> second, ReadOnlySpan<char> third)
         {
             Debug.Assert(first.Length > 0 && second.Length > 0 && third.Length > 0, "should have dealt with empty paths");
+<<<<<<< HEAD
```
Comment (jkotas) [inline src/System.IO.FileSystem/src/System/IO/PathHelpers.cs]: @ahsonkhan Unresolved conflict

**Comment by ahsonkhan** [comment]
Waiting on https://github.com/dotnet/corefx/pull/25961 before this will pass CI.

**Comment by jkotas** [comment]
@dotnet-bot test this please

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test this please

**Pull request merged**